docs/proposals/pod-safety.md

```diff
@@ -0,0 +1,388 @@
+To ensure that shared storage without implicit locking be safe for RWO access, the
+Kubernetes storage subsystem should leverage the strong consistency available through
+the API server and prevent concurrent execution for some types of persistent volumes.
+By leveraging existing concepts, we can allow the scheduler and the kubelet to enforce
+a guarantee that an RWO volume can be used on at-most-one node at a time.
+
+In order to properly support region and zone specific storage, Kubernetes adds node
+selector restrictions to pods derived from the persistent volume. Expanding this
+concept to volume types that have no external metadata to read (NFS, iSCSI) may
+result in adding a label selector to PVs that defines the allowed nodes the storage
+can run on (this is a common requirement for iSCSI, FibreChannel, or NFS clusters).
+
+Because all nodes in a Kubernetes cluster possess a special node name label, it would
+be possible for a controller to observe the scheduling decision of a pod using an
+unsafe volume and "attach" that volume to the node, and also observe the deletion of
+the pod and "detach" the volume from the node. The node would then require that these
+unsafe volumes be "attached" before allowing pod execution. Attach and detach may
+be recorded on the PVC or PV as a new field or materialized via the selection labels.
+
+Possible sequence of operations:
+
+1. Cluster administrator creates a RWO iSCSI persistent volume, available only to
+   nodes with the label selector `storagecluster=iscsi-1`
+2. User requests an RWO volume and is bound to the iSCSI volume
+3. The user creates a pod referencing the PVC
+4. The scheduler observes the pod must schedule on nodes with `storagecluster=iscsi-1`
+   (alternatively this could be enforced in admission) and binds to node `A`
+5. The kubelet on node `A` observes the pod references a PVC that specifies RWO which
+   requires "attach" to be successful
+6. The attach/detach controller observes that a pod has been bound with a PVC that
+   requires "attach", and attempts to execute a CAS update on the PVC/PV attaching
+   it to node `A` and pod 1
+7. The kubelet observes the attach of the PVC/PV and executes the pod
+8. The user terminates the pod
+9. The user creates a new pod that references the PVC
+10. The scheduler binds this new pod to node `B`, which also has `storagecluster=iscsi-1`
+11. The kubelet on node `B` observes the new pod, but sees that the PVC/PV is bound
+    to node `A` and so must wait for detach
+12. The kubelet on node `A` completes the deletion of pod 1
+13. The attach/detach controller observes the first pod has been deleted and that the
+    previous attach of the volume to pod 1 is no longer valid - it performs a CAS
+    update on the PVC/PV clearing its attach state.
+14. The attach/detach controller observes the second pod has been scheduled and
+    attaches it to node `B` and pod 2
+15. The kubelet on node `B` observes the attach and allows the pod to execute.
```

Comment:
I think we already have this flow working on cloudprovider backed storage, we just need to:    
1. teach *all* storage plugins (even those that are not "attachable") to honor it
2. communicate between scheduler and volume controller about where exactly to attach (that's the first half of https://github.com/kubernetes/kubernetes/pull/30044)

The existing flow uses 2 fields in node.Status:
  * volumesInUse: To block volumecontroller from detaching a mounted volume
  * volumesAttached: To block another node from attaching a pre-attached volume

I can think of 2 things we might need to fix off the top of my head:    

1. I think there's a gap with the storage plugin on the node, where some allow 2 pods to bind mount the same volume RWO. I think gce allows this currently.
2. Today, some volume controllers time out and force detach after a "long" time (2h). I think this needs fixing, especially if we're going to use them as described.

@kubernetes/sig-storage